### ISSUE TITLE
Spikes in PPO policy loss

### ISSUE THREAD

**Issue by lvwerra** [lvwerra/trl#101]
We sometimes experience huge loss spikes in the policy loss which either cause the training to fail or take a very long time to recover. It would be useful to investigate where they come from and how to mitigate them. cc @natolambert 

<img width="430" alt="Screenshot 2023-01-24 at 16 24 36" src="https://user-images.githubusercontent.com/8264887/214341251-ce24e511-89c1-49a2-b7e1-8d6e9abb8294.png">

<img width="1294" alt="Screenshot 2023-01-24 at 16 48 34" src="https://user-images.githubusercontent.com/8264887/214341267-287588c1-cda0-4fc4-80f0-5dddd5bfa712.png">

**Comment by younesbelkada** [comment]
One idea could be that we don't mask out the logits corresponding to padding tokens, it is something I am having a look in https://github.com/lvwerra/trl/pull/100 - But I am not sure here if this is really the rootcause of this

**Comment by natolambert** [comment]
Yeah, so something weird is going one with a simultaneous large drop in entropy, clip fraction, etc. Can we log the model outputs at that step? Is there any chance the model output gets stuck on something?

**Comment by DaehanKim** [comment]
I corrected parsing pipeline output and loss spike still remains in sentiment-control notebook example. 
so there may be another reaseon for this unstability.

![image](https://user-images.githubusercontent.com/20675681/216659430-e828a210-203d-4b1e-a3c9-556653795ebb.png)

**Comment by lvwerra** [comment]
Thanks @DaehanKim, yes there is an issue besides the order of the logits. I tracked it down to some changes done in #80 (no spikes at the beginning of the PR and spikes at time of merge) and I started tracking the issue down in #126. I'll report as well here if I figure it out!

**Comment by lvwerra** [comment]
The issue with the loss spikes in the sentiment control notebook was that sometimes only a few new tokens would be generated (1-2) and this would cause the loss to spike. Not sure, yet, where exactly this behaviour comes from but we now know where to look: we can actively generate short sequences and investigate what causes the loss explosion.

**Comment by tengxiaoliu** [comment]
I also experienced the spike loss in my case. I'm using the seq2seq t5 model as the backbone. The model is initialized with a supervised finetuned model. I find that the spike loss comes from steps that have a negative advantage and an extremely high ratio r(\theta). This falls in the situation 6 in the [figure](https://huggingface.co/blog/deep-rl-ppo) below.
<img width="883" alt="image" src="https://user-images.githubusercontent.com/61669825/217460022-1565f798-eace-4c8f-8f49-15af3801ec76.png">

In my case, removing pg_losses1 and only keeping the clipped pg_losses2 can help restrict the ratio and stabilize the loss. I didn't train the model from scratch, so the clip fraction is low (less than 3%). But this is a problem if the clip fraction is too high and most of the loss is clipped. It's not a general solution though, just some findings from my case.

**Comment by b11z** [comment]
It looks like the issue that @DaehanKim noticed is also present in the [gpt2-sentiment.ipynb example](https://github.com/huggingface/trl/blob/main/examples/notebooks/gpt2-sentiment.ipynb). It might be nice to propagate the `extract_pipe_output` fix to that notebook as well.
